units/event_log_tests.cc

```diff
@@ -0,0 +1,213 @@
+#include <cstddef>
+#include <cstdint>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace {
+// Set of supported value types for an attribute.
+enum ValueType { String, Int64, VectorInt64 };
+// Specifies the importance of an Event. Events are logged based on their level
+// of importance. Eg., compile_time.csv only contains the important compile time
+// events while event.log contains all the compile time events.
+enum LogLevel { Low, Medium, High };
+
+// Attribute class
+//
+// An event consists of a set of attributes. Each attribute has a name that
+// indicates what the attribute is (e.g., timestamp, hash, etc) and a value.
+// `Attribute` is the base class for an attribute. For each `ValueType`, there
+// should be an implmentation of `Attribute`. Each implementation must expose an
+// `id_` to indicate which ValueType it is implementing.
+class Attribute {
+ public:
+  // Checks if the `Attribute` is an instance of the given class. Used for
+  // safely casting an `Attribute` to a proper derived class.
+  template <class T>
+  bool isa() {
+    return value_type_ == T::id_;
+  }
+
+  // Safely casts the Attribute to the given class (T). Valid typenames should
+  // be derived from Attribute and have an ValueType `id_` field.
+  template <class T>
+  const T *cast() {
+    return isa<T>() ? static_cast<T *>(this) : nullptr;
+  };
+
+  const char *GetName() const { return name_; }
+  ValueType GetValueType() const { return value_type_; }
```

Comment:
Use `const ValueType& GetValueType() const` to avoid unnecessary copies

Reply:
Since the ValueType's size was 4 bytes, I assumed the copy would be cheap. Sure, I'll change it.

Reply:
Ohhh, I got it wrong, I though that this returns the underlying value. Please disregard then.